modules/loader/source/relocations.cpp

```diff
@@ -645,88 +650,261 @@ bool resolveAArch64(const loader::Relocation &r, loader::ElfMap &map,
   }
   return true;
 }
+
+// Sets the immediate field of an I-type instruction [31:20]
+void writeITypeImm(uint32_t imm, uint8_t *insn_addr) {
+  uint32_t insn;
+  cargo::read_little_endian(&insn, insn_addr);
+  insn = setBitRange(insn, imm, 20, 12);
+  cargo::write_little_endian(insn, insn_addr);
+}
+
+// Sets the immediate field of an U-type instruction [31:12]
+void writeUTypeImm(uint32_t imm, uint8_t *insn_addr) {
+  uint32_t insn;
+  cargo::read_little_endian(&insn, insn_addr);
+  insn = setBitRange(insn, imm, 12, 20);
+  cargo::write_little_endian(insn, insn_addr);
+}
+
+uint32_t getLo12(uint64_t addr) {
+  return static_cast<uint32_t>(getBitRange(addr, 0, 12));
+}
+
+std::optional<uint32_t> getHi20(int64_t addr) {
+  auto signExtend32 = [](int64_t val) {
+    return static_cast<int64_t>(val << 32) >> 32;
+  };
+  addr += 0x800;
+  if (addr != signExtend32(addr)) {
+    (void)fprintf(
+        stderr,
+        "Value %" PRId64 " is out of the range of a 32-bit immediate.\n", addr);
+    return std::nullopt;
+  }
+  return static_cast<uint32_t>(getBitRange(addr, 12, 20));
+}
+
+// assumes little-endian RISCV
+bool resolveRISCV(const loader::Relocation &r, loader::ElfMap &map,
+                  const std::vector<loader::Relocation> &relocations) {
+  const auto relocation_data = decomposeRelocation<uint64_t>(r, map);
+  if (!relocation_data) {
+    return false;
+  }
+  auto [relocation_address, relocation_target_address, symbol_target_address] =
+      *relocation_data;
+
+  // A PC-relative offset (S + A - P)
+  const uint64_t relative_value =
+      symbol_target_address - relocation_target_address;
+
+  auto getPCRelHi20 = [&relocations](
+                          const loader::Relocation &r,
+                          int64_t offset) -> std::optional<loader::Relocation> {
+    auto label_offset = r.offset + offset;
+    // Search for the matching relocation in the same section, with the correct
+    // offset and the PCREL_HI20 type.
+    for (auto &other_r : relocations) {
+      if (r.section_index == other_r.section_index &&
+          label_offset == other_r.offset &&
+          other_r.type == loader::RelocationTypes::RISCV::R_RISCV_PCREL_HI20) {
+        return other_r;
+      }
+    }
+    return std::nullopt;
+  };
+
+  using namespace loader::RelocationTypes::RISCV;
+  switch (r.type) {
+    default: {
+#ifndef NDEBUG
+      auto symbol_name =
+          map.getSymbolName(r.symbol_index).value_or("<unknown symbol>");
+      (void)fprintf(stderr, "Missing symbol: %.*s\n", (int)symbol_name.size(),
+                    symbol_name.data());
+      (void)fprintf(stderr, "   with relocation type: %u\n", r.type);
+#endif
+      CARGO_ASSERT(0, "Unsupported relocation type.");
+      return false;
+    }
+    case R_RISCV_64:
+      // 64-bit relocation: S + A
+      cargo::write_little_endian(symbol_target_address & 0xFFFFFFFF,
```

Comment:
Is symbol_target_address  not a 64 bit value - does it not need to be cast to 32 bit to pick the right little_endian template?

Reply:
Yes you're right, surprised that worked locally.

Reply:
should be fixed now

Reply:
I think it works locally because of little endian, and you right the low half first, it's just that scribbles past on the second write.